### PR TITLE
Allow network-get to return hostnames where previously IPaddr were expected

### LINKED ISSUE #818: traefik errors on cos layer with error: hook failed: "metrics-endpoint-relation-joined"
### Bug Description

Solutions QA team has two runs in which traefik errors on cos layer with error: hook failed: "metrics-endpoint-relation-joined"

The cos layer is built on top of kubernetes-aws. 

From the logs: 

2022-08-21 02:20:42 ERROR juju-log metrics-endpoint:10: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 678, in <module>
    main(TraefikIngressCharm)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/main.py", line 431, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/main.py", line 142, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/framework.py", line 316, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/framework.py", line 784, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/framework.py", line 857, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-0/charm/lib/charms/prometheus_k8s/v0/prometheus_scrape.py", line 1545, in _set_scrape_job_spec
    self._set_unit_ip(event)
  File "/var/lib/juju/agents/unit-traefik-0/charm/lib/charms/prometheus_k8s/v0/prometheus_scrape.py", line 1576, in _set_unit_ip
    unit_ip = str(self._charm.model.get_binding(relation).network.bind_address)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/model.py", line 679, in network
    self._network = self._network_get(self.name, self._relation_id)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/model.py", line 672, in _network_get
    return Network(self._backend.network_get(name, relation_id))
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/model.py", line 724, in __init__
    self.ingress_addresses.append(ipaddress.ip_address(address))
  File "/usr/lib/python3.8/ipaddress.py", line 53, in ip_address
    raise ValueError('%r does not appear to be an IPv4 or IPv6 address' %
ValueError: 'acb396cf5563d429f9ebc8aa23ae47ed-1516332866.us-east-1.elb.amazonaws.com' does not appear to be an IPv4 or IPv6 address

Failed runs: 

https://solutions.qa.canonical.com/testruns/testRun/0b35ee4c-efcf-4f50-a393-696803b31ac9
https://solutions.qa.canonical.com/testruns/testRun/ef230601-4af3-4d6f-89af-4e578985e666

Logs are found on the bottom on the page, on the artifacts repository. 

### To Reproduce

These errors were found on the  fkb-master-kubernetes-focal-aws SKU of our automated test suite.

### Environment

charmed-kubernetes  1.24
stable channel for cos charms. 

### Relevant log output

```shell
Traceback (most recent call last):
  File "./src/charm.py", line 678, in <module>
    main(TraefikIngressCharm)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/main.py", line 431, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/main.py", line 142, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/framework.py", line 316, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/framework.py", line 784, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/framework.py", line 857, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-0/charm/lib/charms/prometheus_k8s/v0/prometheus_scrape.py", line 1545, in _set_scrape_job_spec
    self._set_unit_ip(event)
  File "/var/lib/juju/agents/unit-traefik-0/charm/lib/charms/prometheus_k8s/v0/prometheus_scrape.py", line 1576, in _set_unit_ip
    unit_ip = str(self._charm.model.get_binding(relation).network.bind_address)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/model.py", line 679, in network
    self._network = self._network_get(self.name, self._relation_id)
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/model.py", line 672, in _network_get
    return Network(self._backend.network_get(name, relation_id))
  File "/var/lib/juju/agents/unit-traefik-0/charm/venv/ops/model.py", line 724, in __init__
    self.ingress_addresses.append(ipaddress.ip_address(address))
  File "/usr/lib/python3.8/ipaddress.py", line 53, in ip_address
    raise ValueError('%r does not appear to be an IPv4 or IPv6 address' %
ValueError: 'acb396cf5563d429f9ebc8aa23ae47ed-1516332866.us-east-1.elb.amazonaws.com' does not appear to be an IPv4 or IPv6 address
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PietroPasotti**
Fixes #818

## Checklist

 - [x] Have any types changed? If so, have the type annotations been updated?
 - [x] Do error messages, if any, present charm authors or operators with enough information to solve the problem?

## QA steps

There are no tests for network object properties; ~~we probably should add some~~ --> I added some.

## Documentation changes

We should document somewhere that if juju fails to resolve a hostname, some network-get fields will not be IPs but the 'raw' hostname itself.

## Changelog

- OF won't break if hook tools return hostnames where IPs were previously expected.

### CODE REVIEW

**Comment by rwcarlsen** [comment]
Are we sure we want the union return type?  Or would it possibly make sense to just validate the ip (and log/note if it's not an ip) and then just always return str?  This seems like it would make it easier for user code to handle - only one return type to deal with rather than isinstance branching (or sloppily ignoring the case where it isn't an ip).

**Review by pengale — CHANGES_REQUESTED** [review]
This needs some tests :-)

But the code makes sense, and I like the "try and log on fail" approach to casting to an IP.

**Comment by PietroPasotti** [comment]
> Are we sure we want the union return type? Or would it possibly make sense to just validate the ip (and log/note if it's not an ip) and then just always return str? This seems like it would make it easier for user code to handle - only one return type to deal with rather than isinstance branching (or sloppily ignoring the case where it isn't an ip).

I think I'd prefer that approach as well, but I'd be reluctant to just break any existing code that's instance-checking the address or using some IPaddress attribute without checking the type first.

that looks like a separate issue/PR. 
Will work out some tests for this

**Comment by PietroPasotti** [comment]
Added a test for non-ip-like ingress-address return values. Also verified that the test fails with the expected error message in main.

File: ops/model.py [hunk 1]
```diff
@@ -741,7 +752,7 @@ def bind_address(self) -> Optional['_IPAddress']:
             return None
 
     @property
-    def ingress_address(self):
+    def ingress_address(self) -> Optional['_NetworkAddress']:
```
Comment (rwcarlsen) [inline ops/model.py:755]: Since charmers will likely have only ever written code that expects this to return an ip address object (not a str), they already have broken code.  But they probably have written code that handles this returning None.  So would it be safer for us to log maybe an error - and then return none from here if the cast fails?  This basically would divert the error case down code paths charms already have established and not introduce any new error modes.  The danger is that there _is_ some charm that can handle a string and wants it despite not being able to resolve it to an ip.

**Review by rwcarlsen — COMMENTED** [review]
Just one question/idea.  Once we settle that, I'm +1 on this.

File: ops/model.py (hunk 1 above)
Comment (pengale) [reply ops/model.py:755]: > Since charmers will likely have only ever written code that expects this to return an ip address object (not a str), they already have broken code. But they probably have written code that handles this returning None. So would it be safer for us to log maybe an error - and then return none from here if the cast fails? This basically would divert the error case down code paths charms already have established and not introduce any new error modes. The danger is that there _is_ some charm that can handle a string and wants it despite not being able to resolve it to an ip.

IPs and hostnames are pretty much equivalents of each other in a lot of circumstances. They might lead to branching codepaths once you get to the level of the OS's networking stack. But if I just want to make an http call to an API, my code doesn't need to care about whether or not the string representing the API's address is an IP or a hostname.

In other words, I think that we do risk breaking working charms by being strict about it.

Or to put it another way: the object isn't actually an IP address. It's the host part of the URI, and that can be a set or numbers _or_ a string of letters.

Comment (rwcarlsen) [reply ops/model.py:755]: I mostly agree with that.  But will charm code have been written treating the return object explicitly like an ip_address object?  If so, then returning a string very well could break charm code anyway.  But maybe that's just a pill we need to swallow here.

**Review by pengale — APPROVED** [review]
Yay tests! :-)

File: ops/model.py (hunk 1 above)
Comment (pengale) [reply ops/model.py:755]: It feels most "correct" to me to accept either an ip address or a hostname. I'm not sure what choices charmers have made.

But I don't think that this change breaks them any more than they'd already be broken, if they are breaking on things that aren't ip addresses.

Comment (rwcarlsen) [reply ops/model.py:755]: Okay - I'm convinced that this isn't more bad than the alternatives :P

**Review by rwcarlsen — APPROVED** [review]
LGTM
